### PR TITLE
support of native Windows complation under Visual Studio

### PR DESCRIPTION
**Pull request by allcreater**
### What does this PR do?

Introduce some fixes allowing to build (not-standard compiler features) and run (compiler-specific asserts at STL) program with MSVC.

### Closes Issue(s)

game/animvalue.hh : added include required by std::min/std::max
game/audio.cc : removed dereferencing of end iterator (UB)
game/controllers-midi.cc : added include required by std::stringstream
game/glshader.cc : replaced variable-sized arrays to std::string to remove compiler extension requirement
game/libxml++-impl.hh
game/notes.hh : comparsion functor should not lead to result when a > b and b > a at same time (runtime-checked by compiler at debug build)
game/screen_audiodevices.cc : fixed access beyond array bounds (UB)

### Motivation

Desire to develop and debug this great game under some developer's favourite IDE

### More

- [ ] Added/updated documentation

### Additional Notes

game/screen_audiodevices.cc also fixes strange bug (used device selectons losts after audio device menu reopening) but  don't sure it is it correct. To be honest I don't understand what  is idx and index values at all - so please say what could I do with it.

Sorry about so complex request :'(

### CODE REVIEW

**Comment by Tronic** [comment]
Nice cleanup; much of the code was written for C++98 and clearly some of that still remained after all our efforts of modernising it.

The patch looks good to me.

**Comment by Tronic** [comment]
FWIW, Python somehow manages to Unicode decode any pathnames in a portable manner. Surrogate replacements are used for bytes that are invalid Unicode in underlying names, and that *just works* both ways and everywhere even when files do have unrecognised bytes, or in case of Windows, I suppose, unmapped 16 bit codepoints.

**Comment by allcreater** [comment]
Right, I'm not sure about last commit. 

> Can't we make this work using the UnicodeUtil class functions?
I tried once but never got it works, probably done something wrong. I think I could try again...

Nonetheless I think the issue with encoding is pretty important to discuss it in more detail, but don't know where. Is "issues" right place for it?

Let me tell quickly about my specific issue: Windows 10, system locale is English, paths with Cyrillic symbols turns into "?" at logs (as a result of path.string() calls) and could not be back-converted from song files into path correctly (songparser.cc : 134). Performous not just could opening the files to parse but also writes question marks at logs. This commit _somehow_ fixes it and even works both with MSVC and MXE cross-compilation.
But yes, unfortunately it looks like just workaround and not seems like full-fledged fix and even could works only for me and brokes something for others because we changing the global state. So how about to revert it and return to basic proposal? :) 

PS:
I'm inexperienced contributor and could use some of Github tools incorrectly, so I'm sorry in advance if it will happen :) 

Thank you very much for responses!

**Comment by allcreater** [comment]
Uhm, maybe force resetting (to original proposal commits) was not a great idea, some of discussion context was lost. Probably it's better to do classic revert. I'm sorry.

**Comment by Tronic** [comment]
I don't use Windows and can't really look closer into it -- but hopefully this helps:

Windows provides 8 bit filenames for compatibility with standard C and C++ functions but in presence of non-ASCII filenames these are in no way reliable, and generally you should never use char filenames when developing for Windows. Filenames with Unicode characters would get lost in such a way that the file cannot even be accessed using the char string reported by OS.

- If you get � (U+FFFD unicode replacement character), it means that the original character was not recognised by Unicode conversion working in **replace errors** mode. In case of Cyrillic characters, in principle those should become Unicode Cyrillic.

- If you are seeing ? (ASCII question mark), then it was probably destroyed by Windows itself converting the name into a char string (never use 8 bit filenames on Windows, it just doesn't work).

So, I am hoping that someone already implemented all file I/O using **wchar** filenames (16 bit characters) when running on Windows. These will then need to be converted from UTF-16LE to UTF-8 std::strings for any Performous operations such as logging and UI, although on Windows the logger might also need to print its output in another way because classic Windows terminals don't support UTF-8 (I gather that this is changing now with https://github.com/microsoft/terminal and infolog.txt should in any case be fine with UTF-8).

Then there is the Unicode **surrogate escape** mode, used in Python, because (particularly on UNIX systems) filenames may contain byte sequences that are not valid UTF-8, that are then represented with codepoints U+DC80..U+DCFF (corresponding to chars \x80..\xFF). If the string is converted back into a filename, the exact same bytes get restored (this binary safe encoding is also called UTF-8B). https://www.python.org/dev/peps/pep-0383/ -- but AFAIK there is no similar encoding to handle sequences of invalid UTF-16 such that the original wchars could be restored.

**Comment by Baklap4** [comment]
I'll look into this tonight, looks good as for now. I noticed you added the `.vs` folder to the gitignore. Do you have any launch profiles or some settings regarding visual studio which you might wanna share?

**Comment by Lord-Kamina** [comment]
> Uhm, maybe force resetting (to original proposal commits) was not a great idea, some of discussion context was lost. Probably it's better to do classic revert. I'm sorry.

No need to revert it yet. We'll just take a bit longer to discuss and merge this until we find a better solution IMO. 

Do you have Unicode support enabled in your Windows? (It was a beta feature added about a year or two ago to Windows 10 in an update)

**Comment by allcreater** [comment]
>     * If you are seeing ? (ASCII question mark), then it was probably destroyed by Windows itself converting the name into a char string (never use 8 bit filenames on Windows, it just doesn't work).

Yes, this case :) But it seems it's not a Windows, because fs::path contains right data that coonverted into non-unicode ANSI string with .string() method. My initial proposal was to set special codecvt to do right conversion. Of course it's possible to get path's native string and then convert it manually (e.g. via UnicodeUtils), but it seems we have to do some preprocessor's check - as far as I understand Unix lives with utf8 by default and type of native string should be char, not wchar as under Windows. But it's just assumption.

However...


> Do you have Unicode support enabled in your Windows? (It was a beta feature added about a year or two ago to Windows 10 in an update)

Huh, I just enabled it yesterday with different motivation and never thought of it in this context. But it seems it works! Both logs and file access is working now as with that strange trick in code...

And yes, I have some additional fixes. Partially it's experimental and partially could be useful. Could you advice, please, should I add it to this request or it's better to wait?

**Comment by allcreater** [comment]
And yes, so sorry about too many unlinked changes at one request: I testing the game with friends on big library of songs with attached debugger, so that fixing becomes essential continuation of playing :) Later I hope I will be able to keep separate feature branches.

**Comment by Baklap4** [comment]
@allcreater what did you do to resolve the linker issues? I'm getting them in Release and Debug builds.

**Comment by allcreater** [comment]
@Baklap4 Judjing by linker errors you're almost done :) The last part is to set `-DBUILD_SHARED_LIBS:BOOL="True"`

I don't even try to build Performous with webserver support (so we probably have to install additional libraries with **vcpkg** and set additional variables). Let me try to test it at my configuratuion:)

**Comment by Baklap4** [comment]
for the webserver we need to add `cpprestsdk:x64-windows`. I'll try building with the shared libs on :), thanks!

**Comment by allcreater** [comment]
@Baklap4 yes, but...

Then we got compiler errors and see (at cpprest/details/basic_types.h): 
`#ifndef _WIN32
typedef uint32_t HRESULT; // Needed for PPLX
#endif

#ifdef _UTF16_STRINGS
//
// On Windows, all strings are wide
//
typedef wchar_t char_t;
typedef std::wstring string_t;
#define _XPLATSTR(x) L##x
typedef std::wostringstream ostringstream_t;
typedef std::wofstream ofstream_t;
typedef std::wostream ostream_t;
typedef std::wistream istream_t;
typedef std::wifstream ifstream_t;
typedef std::wistringstream istringstream_t;
typedef std::wstringstream stringstream_t;
#define ucout std::wcout
#define ucin std::wcin
#define ucerr std::wcerr
#else` 

Uhm... Looks like CppRestSDK does not allow to use UTF-8 under Windows... Just don't know how to live with it :/

**Comment by Baklap4** [comment]
Yep that was the last step!
I'll try to get the webserver to build aswell tonight 👍 
Then will test it with my 30k-ish songs.

As for the **debug** build did you also have to add the specific libraries?

**Comment by Baklap4** [comment]
The debug build (x64) got me some linker errors aswell:
```
Severity	Code	Description	Project	File	Line	Suppression State
Error	LNK1319	50 mismatches detected	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\performous-win	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\Performous.exe	1	
Error	LNK2038	mismatch detected for '_ITERATOR_DEBUG_LEVEL': value '0' doesn't match value '2' in 3dobject.cc.obj	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\performous-win	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\xml++.lib(node.cc.obj)	1	
Error	LNK2038	mismatch detected for 'RuntimeLibrary': value 'MD_DynamicRelease' doesn't match value 'MDd_DynamicDebug' in 3dobject.cc.obj	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\performous-win	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\xml++.lib(node.cc.obj)	1	
Error	LNK2038	mismatch detected for '_ITERATOR_DEBUG_LEVEL': value '0' doesn't match value '2' in 3dobject.cc.obj	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\performous-win	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\xml++.lib(attribute.cc.obj)	1	
Error	LNK2038	mismatch detected for 'RuntimeLibrary': value 'MD_DynamicRelease' doesn't match value 'MDd_DynamicDebug' in 3dobject.cc.obj	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\performous-win	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\xml++.lib(attribute.cc.obj)	1	
Error	LNK2038	mismatch detected for '_ITERATOR_DEBUG_LEVEL': value '0' doesn't match value '2' in 3dobject.cc.obj	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\performous-win	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\xml++.lib(contentnode.cc.obj)	1	
Error	LNK2038	mismatch detected for 'RuntimeLibrary': value 'MD_DynamicRelease' doesn't match value 'MDd_DynamicDebug' in 3dobject.cc.obj	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\performous-win	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\xml++.lib(contentnode.cc.obj)	1	
Error	LNK2038	mismatch detected for '_ITERATOR_DEBUG_LEVEL': value '0' doesn't match value '2' in 3dobject.cc.obj	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\performous-win	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\xml++.lib(element.cc.obj)	1	
Error	LNK2038	mismatch detected for 'RuntimeLibrary': value 'MD_DynamicRelease' doesn't match value 'MDd_DynamicDebug' in 3dobject.cc.obj	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\performous-win	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\xml++.lib(element.cc.obj)	1	
Error	LNK2038	mismatch detected for '_ITERATOR_DEBUG_LEVEL': value '0' doesn't match value '2' in 3dobject.cc.obj	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\performous-win	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\xml++.lib(document.cc.obj)	1	
Error	LNK2038	mismatch detected for 'RuntimeLibrary': value 'MD_DynamicRelease' doesn't match value 'MDd_DynamicDebug' in 3dobject.cc.obj	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\performous-win	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\xml++.lib(document.cc.obj)	1	
Error	LNK2038	mismatch detected for '_ITERATOR_DEBUG_LEVEL': value '0' doesn't match value '2' in 3dobject.cc.obj	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\performous-win	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\xml++.lib(domparser.cc.obj)	1	
Error	LNK2038	mismatch detected for 'RuntimeLibrary': value 'MD_DynamicRelease' doesn't match value 'MDd_DynamicDebug' in 3dobject.cc.obj	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\performous-win	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\xml++.lib(domparser.cc.obj)	1	
Error	LNK2038	mismatch detected for '_ITERATOR_DEBUG_LEVEL': value '0' doesn't match value '2' in 3dobject.cc.obj	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\performous-win	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\xml++.lib(parser.cc.obj)	1	
Error	LNK2038	mismatch detected for 'RuntimeLibrary': value 'MD_DynamicRelease' doesn't match value 'MDd_DynamicDebug' in 3dobject.cc.obj	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\performous-win	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\xml++.lib(parser.cc.obj)	1	
Error	LNK2038	mismatch detected for '_ITERATOR_DEBUG_LEVEL': value '0' doesn't match value '2' in 3dobject.cc.obj	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\performous-win	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\xml++.lib(exception.cc.obj)	1	
Error	LNK2038	mismatch detected for 'RuntimeLibrary': value 'MD_DynamicRelease' doesn't match value 'MDd_DynamicDebug' in 3dobject.cc.obj	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\performous-win	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\xml++.lib(exception.cc.obj)	1	
Error	LNK2038	mismatch detected for '_ITERATOR_DEBUG_LEVEL': value '0' doesn't match value '2' in 3dobject.cc.obj	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\performous-win	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\xml++.lib(commentnode.cc.obj)	1	
Error	LNK2038	mismatch detected for 'RuntimeLibrary': value 'MD_DynamicRelease' doesn't match value 'MDd_DynamicDebug' in 3dobject.cc.obj	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\performous-win	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\xml++.lib(commentnode.cc.obj)	1	
Error	LNK2038	mismatch detected for '_ITERATOR_DEBUG_LEVEL': value '0' doesn't match value '2' in 3dobject.cc.obj	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\performous-win	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\xml++.lib(cdatanode.cc.obj)	1	
Error	LNK2038	mismatch detected for 'RuntimeLibrary': value 'MD_DynamicRelease' doesn't match value 'MDd_DynamicDebug' in 3dobject.cc.obj	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\performous-win	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\xml++.lib(cdatanode.cc.obj)	1	
Error	LNK2038	mismatch detected for '_ITERATOR_DEBUG_LEVEL': value '0' doesn't match value '2' in 3dobject.cc.obj	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\performous-win	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\xml++.lib(textnode.cc.obj)	1	
Error	LNK2038	mismatch detected for 'RuntimeLibrary': value 'MD_DynamicRelease' doesn't match value 'MDd_DynamicDebug' in 3dobject.cc.obj	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\performous-win	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\xml++.lib(textnode.cc.obj)	1	
Error	LNK2038	mismatch detected for '_ITERATOR_DEBUG_LEVEL': value '0' doesn't match value '2' in 3dobject.cc.obj	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\performous-win	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\xml++.lib(processinginstructionnode.cc.obj)	1	
Error	LNK2038	mismatch detected for 'RuntimeLibrary': value 'MD_DynamicRelease' doesn't match value 'MDd_DynamicDebug' in 3dobject.cc.obj	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\performous-win	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\xml++.lib(processinginstructionnode.cc.obj)	1	
Error	LNK2038	mismatch detected for '_ITERATOR_DEBUG_LEVEL': value '0' doesn't match value '2' in 3dobject.cc.obj	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\performous-win	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\xml++.lib(entityreference.cc.obj)	1	
Error	LNK2038	mismatch detected for 'RuntimeLibrary': value 'MD_DynamicRelease' doesn't match value 'MDd_DynamicDebug' in 3dobject.cc.obj	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\performous-win	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\xml++.lib(entityreference.cc.obj)	1	
Error	LNK2038	mismatch detected for '_ITERATOR_DEBUG_LEVEL': value '0' doesn't match value '2' in 3dobject.cc.obj	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\performous-win	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\xml++.lib(entitydeclaration.cc.obj)	1	
Error	LNK2038	mismatch detected for 'RuntimeLibrary': value 'MD_DynamicRelease' doesn't match value 'MDd_DynamicDebug' in 3dobject.cc.obj	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\performous-win	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\xml++.lib(entitydeclaration.cc.obj)	1	
Error	LNK2038	mismatch detected for '_ITERATOR_DEBUG_LEVEL': value '0' doesn't match value '2' in 3dobject.cc.obj	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\performous-win	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\xml++.lib(xincludestart.cc.obj)	1	
Error	LNK2038	mismatch detected for 'RuntimeLibrary': value 'MD_DynamicRelease' doesn't match value 'MDd_DynamicDebug' in 3dobject.cc.obj	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\performous-win	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\xml++.lib(xincludestart.cc.obj)	1	
Error	LNK2038	mismatch detected for '_ITERATOR_DEBUG_LEVEL': value '0' doesn't match value '2' in 3dobject.cc.obj	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\performous-win	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\xml++.lib(xincludeend.cc.obj)	1	
Error	LNK2038	mismatch detected for 'RuntimeLibrary': value 'MD_DynamicRelease' doesn't match value 'MDd_DynamicDebug' in 3dobject.cc.obj	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\performous-win	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\xml++.lib(xincludeend.cc.obj)	1	
Error	LNK2038	mismatch detected for '_ITERATOR_DEBUG_LEVEL': value '0' doesn't match value '2' in 3dobject.cc.obj	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\performous-win	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\xml++.lib(internal_error.cc.obj)	1	
Error	LNK2038	mismatch detected for 'RuntimeLibrary': value 'MD_DynamicRelease' doesn't match value 'MDd_DynamicDebug' in 3dobject.cc.obj	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\performous-win	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\xml++.lib(internal_error.cc.obj)	1	
Error	LNK2038	mismatch detected for '_ITERATOR_DEBUG_LEVEL': value '0' doesn't match value '2' in 3dobject.cc.obj	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\performous-win	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\xml++.lib(attributedeclaration.cc.obj)	1	
Error	LNK2038	mismatch detected for 'RuntimeLibrary': value 'MD_DynamicRelease' doesn't match value 'MDd_DynamicDebug' in 3dobject.cc.obj	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\performous-win	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\xml++.lib(attributedeclaration.cc.obj)	1	
Error	LNK2038	mismatch detected for '_ITERATOR_DEBUG_LEVEL': value '0' doesn't match value '2' in 3dobject.cc.obj	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\performous-win	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\xml++.lib(attributenode.cc.obj)	1	
Error	LNK2038	mismatch detected for 'RuntimeLibrary': value 'MD_DynamicRelease' doesn't match value 'MDd_DynamicDebug' in 3dobject.cc.obj	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\performous-win	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\xml++.lib(attributenode.cc.obj)	1	
Error	LNK2038	mismatch detected for '_ITERATOR_DEBUG_LEVEL': value '0' doesn't match value '2' in 3dobject.cc.obj	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\performous-win	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\xml++.lib(dtd.cc.obj)	1	
Error	LNK2038	mismatch detected for 'RuntimeLibrary': value 'MD_DynamicRelease' doesn't match value 'MDd_DynamicDebug' in 3dobject.cc.obj	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\performous-win	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\xml++.lib(dtd.cc.obj)	1	
Error	LNK2038	mismatch detected for '_ITERATOR_DEBUG_LEVEL': value '0' doesn't match value '2' in 3dobject.cc.obj	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\performous-win	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\xml++.lib(outputbuffer.cc.obj)	1	
Error	LNK2038	mismatch detected for 'RuntimeLibrary': value 'MD_DynamicRelease' doesn't match value 'MDd_DynamicDebug' in 3dobject.cc.obj	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\performous-win	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\xml++.lib(outputbuffer.cc.obj)	1	
Error	LNK2038	mismatch detected for '_ITERATOR_DEBUG_LEVEL': value '0' doesn't match value '2' in 3dobject.cc.obj	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\performous-win	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\xml++.lib(ostreamoutputbuffer.cc.obj)	1	
Error	LNK2038	mismatch detected for 'RuntimeLibrary': value 'MD_DynamicRelease' doesn't match value 'MDd_DynamicDebug' in 3dobject.cc.obj	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\performous-win	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\xml++.lib(ostreamoutputbuffer.cc.obj)	1	
Error	LNK2038	mismatch detected for '_ITERATOR_DEBUG_LEVEL': value '0' doesn't match value '2' in 3dobject.cc.obj	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\performous-win	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\xml++.lib(parse_error.cc.obj)	1	
Error	LNK2038	mismatch detected for 'RuntimeLibrary': value 'MD_DynamicRelease' doesn't match value 'MDd_DynamicDebug' in 3dobject.cc.obj	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\performous-win	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\xml++.lib(parse_error.cc.obj)	1	
Error	LNK2038	mismatch detected for '_ITERATOR_DEBUG_LEVEL': value '0' doesn't match value '2' in 3dobject.cc.obj	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\performous-win	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\xml++.lib(wrapped_exception.cc.obj)	1	
Error	LNK2038	mismatch detected for 'RuntimeLibrary': value 'MD_DynamicRelease' doesn't match value 'MDd_DynamicDebug' in 3dobject.cc.obj	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\performous-win	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\xml++.lib(wrapped_exception.cc.obj)	1	
Error	LNK2038	mismatch detected for '_ITERATOR_DEBUG_LEVEL': value '0' doesn't match value '2' in 3dobject.cc.obj	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\performous-win	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\xml++.lib(validity_error.cc.obj)	1	
Error	LNK2038	mismatch detected for 'RuntimeLibrary': value 'MD_DynamicRelease' doesn't match value 'MDd_DynamicDebug' in 3dobject.cc.obj	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\performous-win	C:\Users\aspeiard\Desktop\Persoonlijk\performous-win\out\build\x64-Debug\xml++.lib(validity_error.cc.obj)	1	
```

**Comment by allcreater** [comment]
@Baklap4 hmmm, looks very familiar, I'm defiinitely was fixing it somehow, but forgot the method. Let me think a bit :)

About testing on real songs: personally I still got some errors and strange behaviors at some songs, but mostly already workarounded them. If you will experience some specific bugs too, then maybe we could make separate experimental branch?

**Comment by allcreater** [comment]
@Baklap4 oh yes, ensure you don't mix debug and release libraries, debug libraries lives at _vcpkg/installed/x64-windows/debug/lib/_

**Comment by Baklap4** [comment]
Alright i've managed to get a working build (debug + release) on windows. It definitely had something to do with the debug/vs release libs.

I do still notice loads of warnings while building.
Tonight i'll be checking out how to get the webserver support up and running :)! Good work so far!

**Comment by allcreater** [comment]
@Baklap4 yes, warnings worries me too, so I definitely like to start to unrooting them. But i think I'd prefer to do it at some special branch.

**Comment by Baklap4** [comment]
Agreed. Let's first fix this build, then focus on the warnings in a separate pull request

**Comment by Baklap4** [comment]
Patch for the webserver
[patch-webserver.txt](https://github.com/performous/performous/files/4282505/patch-webserver.txt)

On windows there's an issue with urlacl (url access control list?) where if you bind to 0.0.0.0 you either have to be administrator, or you have to add it to your 'reserved' list using:
 `netsh http add urlacl url=http://0.0.0.0:8080/ user=BUILTIN\Users listen=yes`

**Comment by allcreater** [comment]
@Baklap4 applied :)

**Comment by Baklap4** [comment]
probably should need some rework though since there's a lot of conversions going on in my code switching between utf-8 strings and string_t of cpprestsdk but it's a start :)

**Comment by allcreater** [comment]
@Baklap4 personally I digging into very specific bug with duet at some songs (unsorted collection assert at std::merge at song.cc)  and not sure that finish it very fast, so if you plan to patch something else, I'll be glad to do it later :)

By the way, about bug. Do not tell me what could mean the reason of constants 0 and 1 at merging vocal tracks there? Seems like we do not need to do merge at all and just take SongParserUtil::DUET_BOTH, or at least take selected at duet menu tracks... But if the question is not easy, don't want to disturb you anyway :)

**Comment by Baklap4** [comment]
@allcreater are you within the discord of performous? Feel free to join there to do some chatting (i'm Asgard Sings!) :)! 

At what line of song.cc do you have this bug?

**Comment by Lord-Kamina** [comment]
> Seems like we do not need to do merge at all and just take SongParserUtil::DUET_BOTH, or at least take selected at duet menu tracks... But if the question is not easy, don't want to disturb you anyway :)

Not sure what you mean exactly.

That bit of code was my handiwork; and the purpose is to enable performous to automatically create a 1-player version of duet songs; so as to eliminate the need to have more than one  txt tor a song.

**Comment by Baklap4** [comment]
The problem with a duet song is that s1, and s2 are 'unordered' therefore they can't be merged together, atleast that's the exception we're getting

**Comment by allcreater** [comment]
> That bit of code was my handiwork; and the purpose is to enable performous to automatically create a 1-player version of duet songs; so as to eliminate the need to have more than one txt tor a song.

Yes, I understand the purpose, but at least txt files are already merged into one special voice track (at songparser-txt.cc : 48 and below), and this track even processed to remove duplicates and do something with overlapping notes. It stores as SongParserUtil::DUET_BOTH.

Then when we take vocalTracks 0 and 1, we merging not player 1 and player 2's tracks, but actually player 1 and SongParserUtil::DUET_BOTH! Fixed "both" track could be unsorted a bit and in that case debug std::merge version will trigger an assert.

The issue is not so important (in my case only one song triggers that assert) and mostly all works great, but all that things in common looks a bit suspiciously for me :)

**Comment by allcreater** [comment]
@Lord-Kamina just don't sure what type is better. Ok :)

**Comment by Lord-Kamina** [comment]
> @Lord-Kamina just don't sure what type is better. Ok :)

Long works too of course, but most likely you're allocating more memory than is necessary by using it.

**Comment by allcreater** [comment]
Maybe we should also add some overflow checks, but it should not happen in real songs :)

I'd like to fix compiler warning later, but I think it's better to complete this request first :)

**Comment by Baklap4** [comment]
Ello, @allcreater can you sync this up with your remote experimental branch?
And if so what loose ends do we have left?

We need to think of something with using glyphs in a font which doesn't support them
1. we can add another glyph which contains them
2. Find some way around this
3. Stick around with the ?? boxes

Some testing still needs to be done.
Personally I only care for the singstar/karaoke part, but other use it for guitar hero and dance dance revolution aswell so we need to get that tested

aside from those two points we should be ready to merge this to master? or am i still missing stuff?

File: game/screen.hh [hunk 1]
```diff
@@ -140,6 +140,7 @@ private:
 #endif
 };
 
+#if !(_MSC_VER && !__INTEL_COMPILER) //MSVC does not supports redefintions of specialized static member varables
```
Comment (swurzinger) [inline game/screen.hh]: did this change with VS 2019 16.5+?
see https://stackoverflow.com/questions/32321006/resolving-definitions-of-specialized-static-member-variables-of-templated-classe

**Comment by Baklap4** [comment]
So where are we at this? @allcreater do you think there's still stuff to test/check? If not i'd love to get this merged

**Comment by allcreater** [comment]
@Baklap4 OMG, thank you so much! When I was confused what to do with new changes you already fixed it! 

So I'm already trying to get it work, and I'll write the answer as soon as I finish.

Anyway I'm trust your opinion about readyness, so you could merge it when you wish :)

**Comment by Baklap4** [comment]
@Lord-Kamina if https://github.com/performous/performous/pull/507 gets merged then this can be merged. So let's see if we can resolve the issues regarding #507 asap :)

**Comment by Baklap4** [comment]
closing this in favor of #537

File: game/screen.hh (hunk 1 above)
Comment (Baklap4) [reply game/screen.hh:143]: Item has been moved to another pR (#537 )

**Pull request closed without merging**